File: backend/utils/storage_paths.py
"""
Supabase Storage Path Utilities
Standardized path convention for all storage operations
"""

import uuid
from datetime import datetime
from typing import Optional


def generate_workspace_path(workspace_slug: str, category: str, filename: Optional[str] = None) -> str:
    """
    Generate standardized storage path for workspace files
    
    Format: workspace/<slug>/<category>/<YYYY>/<MM>/<DD>/<uuid>[-<filename>]
    
    Args:
        workspace_slug: Workspace identifier
        category: File category (uploads, exports, backups, assets, temp, logs)
        filename: Original filename (optional)
    
    Returns:
        Standardized storage path
    """
    date_path = datetime.now().strftime("%Y/%m/%d")
    unique_id = str(uuid.uuid4())
    
    if filename:
        # Sanitize filename and append UUID
        safe_filename = filename.replace(" ", "_").replace("/", "_")
        path = f"workspace/{workspace_slug}/{category}/{date_path}/{unique_id}-{safe_filename}"
    else:
        path = f"workspace/{workspace_slug}/{category}/{date_path}/{unique_id}"
    
    return path


def generate_user_path(user_id: str, category: str, filename: Optional[str] = None) -> str:
    """
    Generate standardized storage path for user files
    
    Format: users/<user_id>/<category>/<YYYY>/<MM>/<DD>/<uuid>[-<filename>]
    
    Args:
        user_id: User identifier
        category: File category (avatars, documents)
        filename: Original filename (optional)
    
    Returns:
        Standardized storage path
    """
    date_path = datetime.now().strftime("%Y/%m/%d")
    unique_id = str(uuid.uuid4())
    
    if filename:
        safe_filename = filename.replace(" ", "_").replace("/", "_")
        path = f"users/{user_id}/{category}/{date_path}/{unique_id}-{safe_filename}"
    else:
        path = f"users/{user_id}/{category}/{date_path}/{unique_id}"
    
    return path


def generate_intelligence_path(query_hash: str, filename: Optional[str] = None) -> str:
    """
    Generate standardized storage path for intelligence outputs
    
    Format: intelligence/<query_hash>/<YYYY>/<MM>/<DD>/<uuid>[-<filename>]
    
    Args:
        query_hash: Hash of the query for deduplication
        filename: Original filename (optional)
    
    Returns:
        Standardized storage path
    """
    date_path = datetime.now().strftime("%Y/%m/%d")
    unique_id = str(uuid.uuid4())
    
    if filename:
        safe_filename = filename.replace(" ", "_").replace("/", "_")
        path = f"intelligence/{query_hash}/{date_path}/{unique_id}-{safe_filename}"
    else:
        path = f"intelligence/{query_hash}/{date_path}/{unique_id}"
    
    return path


def parse_storage_path(path: str) -> dict:
    """
    Parse storage path to extract metadata
    
    Args:
        path: Storage path
    
    Returns:
        Dictionary with parsed metadata
    """
    parts = path.split('/')
    
    result = {
        'type': None,
        'owner_id': None,
        'category': None,
        'date_path': None,
        'unique_id': None,
        'filename': None
    }
    
    if len(parts) >= 6:
        result['type'] = parts[0]  # workspace, users, intelligence
        
        if result['type'] == 'workspace':
            result['owner_id'] = parts[1]  # workspace_slug
            result['category'] = parts[2]  # category
            result['date_path'] = "/".join(parts[3:6])  # YYYY/MM/DD
            result['unique_id'] = parts[6].split('-')[0] if len(parts) > 6 else None
            result['filename'] = "-".join(parts[6].split('-')[1:]) if len(parts) > 6 and '-' in parts[6] else None
        elif result['type'] == 'users':
            result['owner_id'] = parts[1]  # user_id
            result['category'] = parts[2]  # category
            result['date_path'] = "/".join(parts[3:6])  # YYYY/MM/DD
            result['unique_id'] = parts[6].split('-')[0] if len(parts) > 6 else None
            result['filename'] = "-".join(parts[6].split('-')[1:]) if len(parts) > 6 and '-' in parts[6] else None
        elif result['type'] == 'intelligence':
            result['owner_id'] = parts[1]  # query_hash
            result['category'] = 'intelligence'
            result['date_path'] = "/".join(parts[2:5])  # YYYY/MM/DD
            result['unique_id'] = parts[5].split('-')[0] if len(parts) > 5 else None
            result['filename'] = "-".join(parts[5].split('-')[1:]) if len(parts) > 5 and '-' in parts[5] else None
    
    return result


# Bucket mapping constants
BUCKET_MAPPING = {
    # Workspace buckets
    'uploads': 'workspace-uploads',
    'exports': 'workspace-exports', 
    'backups': 'workspace-backups',
    'assets': 'workspace-assets',
    'temp': 'workspace-temp',
    'logs': 'workspace-logs',
    
    # User buckets
    'avatars': 'user-avatars',
    'documents': 'user-documents',
    
    # Special buckets
    'intelligence': 'intelligence-vault',
    'exports': 'user-data',
    'gdpr': 'user-data'
}


def get_bucket_for_category(category: str) -> str:
    """
    Get Supabase bucket name for file category
    
    Args:
        category: File category
    
    Returns:
        Supabase bucket name
    """
    return BUCKET_MAPPING.get(category, 'workspace-uploads')

File: src/lib/storage-paths.ts
/**
 * Supabase Storage Path Utilities
 * Standardized path convention for all storage operations
 */

// ============================================================================
// BUCKET CONFIGURATION
// ============================================================================

export const BUCKETS = {
  WORKSPACE_UPLOADS: 'workspace-uploads',
  WORKSPACE_EXPORTS: 'workspace-exports',
  WORKSPACE_BACKUPS: 'workspace-backups',
  WORKSPACE_ASSETS: 'workspace-assets',
  WORKSPACE_LOGS: 'workspace-logs',
  INTELLIGENCE_VAULT: 'intelligence-vault',
  USER_AVATARS: 'user-avatars',
  USER_DOCUMENTS: 'user-documents',
  USER_DATA: 'user-data'
} as const;

// ============================================================================
// PATH GENERATION FUNCTIONS
// ============================================================================

/**
 * Generate standardized workspace storage path
 * Format: workspace/<workspace-slug>/<category>/<YYYY-MM-DD>/<uuid>-<filename>
 */
export function generateWorkspacePath(
  workspaceSlug: string,
  category: string,
  filename: string
): string {
  const date = new Date().toISOString().split('T')[0]; // YYYY-MM-DD
  const uuid = crypto.randomUUID();
  const cleanFilename = filename.replace(/[^a-zA-Z0-9.-]/g, '_');

  return `workspace/${workspaceSlug}/${category}/${date}/${uuid}-${cleanFilename}`;
}

/**
 * Generate standardized user storage path
 * Format: users/<user-id>/<category>/<YYYY-MM-DD>/<uuid>-<filename>
 */
export function generateUserPath(
  userId: string,
  category: string,
  filename: string
): string {
  const date = new Date().toISOString().split('T')[0]; // YYYY-MM-DD
  const uuid = crypto.randomUUID();
  const cleanFilename = filename.replace(/[^a-zA-Z0-9.-]/g, '_');

  return `users/${userId}/${category}/${date}/${uuid}-${cleanFilename}`;
}

/**
 * Generate intelligence vault path
 * Format: intelligence-vault/<workspace-slug>/<YYYY-MM-DD>/<uuid>.json
 */
export function generateIntelligencePath(
  workspaceSlug: string,
  queryId?: string
): string {
  const date = new Date().toISOString().split('T')[0]; // YYYY-MM-DD
  const uuid = queryId || crypto.randomUUID();

  return `intelligence-vault/${workspaceSlug}/${date}/${uuid}.json`;
}

// ============================================================================
// BUCKET SELECTION
// ============================================================================

/**
 * Get appropriate bucket for file category
 */
export function getBucketForCategory(category: string): string {
  switch (category.toLowerCase()) {
    case 'avatar':
    case 'avatars':
      return BUCKETS.USER_AVATARS;

    case 'document':
    case 'documents':
    case 'uploads':
      return BUCKETS.WORKSPACE_UPLOADS;

    case 'export':
    case 'exports':
      return BUCKETS.WORKSPACE_EXPORTS;

    case 'backup':
    case 'backups':
      return BUCKETS.WORKSPACE_BACKUPS;

    case 'asset':
    case 'assets':
      return BUCKETS.WORKSPACE_ASSETS;

    case 'log':
    case 'logs':
      return BUCKETS.WORKSPACE_LOGS;

    case 'intelligence':
    case 'evidence':
      return BUCKETS.INTELLIGENCE_VAULT;

    case 'user-data':
      return BUCKETS.USER_DATA;

    default:
      return BUCKETS.WORKSPACE_UPLOADS;
  }
}

// ============================================================================
// PATH PARSING
// ============================================================================

/**
 * Parse storage path to extract components
 */
export function parseStoragePath(storagePath: string): {
  type: 'workspace' | 'user' | 'intelligence';
  workspaceSlug?: string;
  userId?: string;
  category?: string;
  date?: string;
  filename?: string;
} | null {
  // Workspace path: workspace/<workspace-slug>/<category>/<date>/<filename>
  const workspaceMatch = storagePath.match(/^workspace\/([^/]+)\/([^/]+)\/([^/]+)\/(.+)$/);
  if (workspaceMatch) {
    return {
      type: 'workspace',
      workspaceSlug: workspaceMatch[1],
      category: workspaceMatch[2],
      date: workspaceMatch[3],
      filename: workspaceMatch[4]
    };
  }

  // User path: users/<user-id>/<category>/<date>/<filename>
  const userMatch = storagePath.match(/^users\/([^/]+)\/([^/]+)\/([^/]+)\/(.+)$/);
  if (userMatch) {
    return {
      type: 'user',
      userId: userMatch[1],
      category: userMatch[2],
      date: userMatch[3],
      filename: userMatch[4]
    };
  }

  // Intelligence path: intelligence-vault/<workspace-slug>/<date>/<filename>
  const intelligenceMatch = storagePath.match(/^intelligence-vault\/([^/]+)\/([^/]+)\/(.+)$/);
  if (intelligenceMatch) {
    return {
      type: 'intelligence',
      workspaceSlug: intelligenceMatch[1],
      date: intelligenceMatch[2],
      filename: intelligenceMatch[3]
    };
  }

  return null;
}

// ============================================================================
// VALIDATION
// ============================================================================

/**
 * Validate storage path format
 */
export function validateStoragePath(storagePath: string): boolean {
  const parsed = parseStoragePath(storagePath);
  return parsed !== null;
}

/**
 * Check if path belongs to workspace
 */
export function isWorkspacePath(storagePath: string): boolean {
  const parsed = parseStoragePath(storagePath);
  return parsed?.type === 'workspace';
}

/**
 * Check if path belongs to user
 */
export function isUserPath(storagePath: string): boolean {
  const parsed = parseStoragePath(storagePath);
  return parsed?.type === 'user';
}

/**
 * Check if path belongs to intelligence vault
 */
export function isIntelligencePath(storagePath: string): boolean {
  const parsed = parseStoragePath(storagePath);
  return parsed?.type === 'intelligence';
}

// ============================================================================
// HELPERS
// ============================================================================

/**
 * Extract workspace slug from storage path
 */
export function extractWorkspaceSlug(storagePath: string): string | null {
  const parsed = parseStoragePath(storagePath);
  return parsed?.workspaceSlug || null;
}

/**
 * Extract user ID from storage path
 */
export function extractUserId(storagePath: string): string | null {
  const parsed = parseStoragePath(storagePath);
  return parsed?.userId || null;
}

/**
 * Extract category from storage path
 */
export function extractCategory(storagePath: string): string | null {
  const parsed = parseStoragePath(storagePath);
  return parsed?.category || null;
}

/**
 * Generate public URL for storage path
 */
export function generatePublicUrl(bucket: string, storagePath: string): string {
  const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
  if (!supabaseUrl) {
    throw new Error('NEXT_PUBLIC_SUPABASE_URL not configured');
  }

  return `${supabaseUrl}/storage/v1/object/public/${bucket}/${storagePath}`;
}
